### DISCUSSION TITLE
Usage scenarios of `Streaming command response`

### DISCUSSION THREAD

**Discussion by zzzzrf** [strongswan/strongswan#2070] [Q&A] [answered]
When I read the davici code, I found that in addition to `Invoking commands` and `event handling`, there is also a `Streaming command response`. But I don't understand in what scenario I should use it.

For example: If I want to initiate an IKE connection request, I only need to do this:

```c
static void initiate_cb(struct davici_conn *conn, int err, const char *name,
                            struct davici_response *res, void *user)
{
    davici_dump(res, name, " ", 1, 1, stderr);
}

static int write_callback(struct thread *th)
{
    struct davici_conn *c = (struct davici_conn *)THREAD_ARG(th);
    davici_write(c);

    return 0;
}

int ike_initiate(struct cmd_element *self, struct vty *vty, int argc,
                 char **argv) 
{
    struct davici_request *req;

    davici_new_cmd("initiate", &req);
    davici_kvf(req, "ike", "%s", "h2h");
    davici_queue(vici_conn, req, initiate_cb, vty);
    thread_add_write(master, write_callback, vici_conn, davici_conn_socket(vici_conn));
    return 0;
}
```

But I read the vici related documents and swanctl/commond/initiate.c They all registered `control-log events` before calling `initiate cmd`.

So I'm confused:

1. Why you must register `control-log events`
2. What is the usage scenario of `Streaming command response`?

Thanks for the help

**Comment by tobiasbrunner** [comment, marked as answer]
You might want to read the [README for the vici protocol](https://github.com/strongswan/strongswan/blob/master/src/libcharon/plugins/vici/README.md). It describes what "response streaming" is and which commands generate such events.

For the `initiate()` and `terminate()` commands, vici can stream log messages to the client. If you don't register for the event, you simply won't receive such messages. You could set `loglevel` to `-1` to explicitly indicate that you are not interested in log messages during the call, which avoids the overhead of preparing and attempting to send a vici message for every log message.
